### PR TITLE
Overlord Refactor 1

### PR DESCRIPTION
**Pull request by jfinzel**
Made some major revisions to add the configuration, which I didn't have before, and I tried to make tests that were more exhaustive and more refactored.  Be careful with the bomb!

### CODE REVIEW

File: overlord/spec/features/configuration_validation.feature [hunk 1]
```diff
@@ -0,0 +1,19 @@
+Feature: Valid configuration inputs only accepted
```
Comment (incazteca) [inline overlord/spec/features/configuration_validation.feature]: Merge this file with the other validation file. They both are similar.

**Comment by incazteca** [comment]
Tests are much more thorough this time. However try to keep them organized. Put the features directory in the application's root directory and keep the features files in there. Create a step_definitions directory and place that in the features directory. Keep the spec files in the spec directory.

File: overlord/spec/features/explode_steps_spec.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
+require_relative '../spec_helper.rb'
```
Comment (incazteca) [inline overlord/spec/features/explode_steps_spec.rb]: If this is a spec file place it in the spec directory.

File: overlord/app.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    end
+  end
+
+  def process_code
+    bomb.process_code(params['code'])
+    rescue BombError => error
+      errors << error.message
+  end
+
+  def start_time
```
Comment (incazteca) [inline overlord/app.rb]: I see you have a method to start the timer however have not yet implemented the bomb to go off after a certain time or test its detonation.

**Comment by incazteca** [comment]
@jfinzel Please finish implementation of the bomb's detonation and relevant tests. Also please organize your tests a little better so it's easier to find them.

File: overlord/app.rb (hunk 3 above)
Comment (jfinzel) [inline overlord/app.rb]: The detonation is all working, according to the exercise.  When it's
detonated that is indicated and the buttons don't work.

I didn't do anything with the timer; it's not mentioned as part of the
features so I just let it there so I may just as well remove it.

*Jeremy Finzel*

*Senior Database Developer*
*M* 630-777-4520

File: overlord/spec/spec_helper.rb [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  #     --seed 1234
+  config.order = :random
+
+  # Seed global randomization in this process using the `--seed` CLI option.
+  # Setting this allows you to use `--seed` to deterministically reproduce
+  # test failures related to randomization by passing the same `--seed` value
+  # as the one that triggered the failure.
+  Kernel.srand config.seed
+=end
+end
```
Comment (djkotowski) [inline overlord/spec/spec_helper.rb]: Go ahead and remove generated comments, especially when they build up hound errors.

File: overlord/spec/features/default_control.feature [hunk 5]
```diff
@@ -0,0 +1,29 @@
+Feature: Control the Bomb with Default Codes
+	In order to make sure we can use a bomb safely
+  As a super villain
+  I want to be able to control the bomb with default codes
+
+  Background:
+  	Given I have booted the bomb
+  	And I change no configuration code
+
+  Scenario Outline: Try to control the bomb
```
Comment (djkotowski) [inline overlord/spec/features/default_control.feature]: This seems like many scenarios compounded into one.  I would at least separate activation and deactivation tests.

File: overlord/spec/features/changed_deactivate_control.feature [hunk 6]
```diff
@@ -0,0 +1,27 @@
+Feature: Control the Bomb with Changed Deactivation Code
+	In order to make sure we can use a bomb safely
```
Comment (djkotowski) [inline overlord/spec/features/changed_deactivate_control.feature]: Indentation.

File: overlord/spec/features/boot_bomb.feature [hunk 7]
```diff
@@ -0,0 +1,17 @@
+Feature: Boot the bomb
+  In order to control a bomb
+  As a super villain
+  I want to boot the bomb safely
+
+  Background:
+  	Given I have booted the bomb
+
+  Scenario Outline: Configure without activating
```
Comment (djkotowski) [inline overlord/spec/features/boot_bomb.feature]: This is confusing.  I'd separate this into two scenarios.

File: overlord/spec/bomb_spec.rb [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+  let(:bomb) { Bomb.new }
+
+  it "adds 1 to invalid count when an invalid code is entered", :happy do
+    bomb.deactivate(1111)
+    expect(bomb.invalid_count).to eq(1)
+  end
+
+  it "should not detonate after 1 invalid code is entered", :happy do
+    bomb.deactivate(1111)
+    expect(bomb.detonated?).to eq(false)
```
Comment (djkotowski) [inline overlord/spec/bomb_spec.rb]: Avoid boolean comparisons.
```ruby
expect(bomb).not_to be_detonated
```

File: overlord/views/_configure_form.erb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+<form action="/" method="post">
+	<label for="activate_code">Override activation code: </label>
+  <input id="activate_code" type="text" name="activate_code" autofocus><br>
+  <label for="deactivate_code">Override deactivation code: </label>
+  <input id="deactivate_code" type="text" name="deactivate_code">
+  <input id="submit" type="submit" name="submit">
+</form>
```
Comment (djkotowski) [inline overlord/views/_configure_form.erb]: I like extracting this out to a partial.  The modularity you establish in your views is really good.

File: overlord/spec/features/explode_steps_spec.rb [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+include OverlordTest
+
+When(/^I enter the wrong deactivation code three times$/) do
+  def enter_wrong_code
+    fill_in "code", with: "1111"
+    click_button "submit"
+  end
+  (1..3).each { enter_wrong_code }
```
Comment (djkotowski) [inline overlord/spec/features/explode_steps_spec.rb]: Can also use `3.times { enter_wrong_code }`

File: overlord/spec/features/explode_steps_spec.rb [hunk 11]
```diff
@@ -0,0 +1,18 @@
+require_relative '../spec_helper.rb'
+include OverlordTest
+
+When(/^I enter the wrong deactivation code three times$/) do
+  def enter_wrong_code
```
Comment (djkotowski) [inline overlord/spec/features/explode_steps_spec.rb]: Move this outside the step.

**Comment by djkotowski** [comment]
Your features directory should be outside the spec directory.

File: overlord/spec/features/changed_both_control.feature [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+  As a super villain
+  I want to be able to control the bomb after changing both codes
+
+  Background:
+  	Given I have booted the bomb
+  	And I fill in the activation code as "9385"
+    And I fill in the deactivation code as "3264"
+    And I submit the configuration
+
+  Scenario Outline: Try to control the bomb
```
Comment (djkotowski) [inline overlord/spec/features/changed_both_control.feature]: Too many columns in your table.  This really should be split into multiple scenarios / scenario outlines.

File: overlord/spec/features/changed_deactivate_control.feature [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+  As a super villain
+  I want to be able to control the bomb after changing the deactivation code
+
+  Background:
+  	Given I have booted the bomb
+  	And I change the deactivation code to "3264"
+    And I enter the default activation code 1 time(s)
+    And the status indicator shows as activated
+
+  Scenario Outline: Try to deactivate the bomb
```
Comment (djkotowski) [inline overlord/spec/features/changed_deactivate_control.feature]: This is an example of how to utilize scenario outlines well.  You DRY it out, but the scenario is not too dynamic where it confuses the reader.

File: overlord/spec/features/configure_steps.rb [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+include OverlordTest
+
+When(/^I change no configuration code$/) do
+  click_button "submit"
+end
+
+When(/^I change the activation code to "(.*?)"$/) do |code|
+  fill_in "activate_code", with: code
+  click_button "submit"
+  expect(page.status_code).to eq(200)
```
Comment (djkotowski) [inline overlord/spec/features/configure_steps.rb]: Don't include assertions in your `Given` or `When` statements.  There is no need for them.

**Comment by djkotowski** [comment]
This is looking good.  Nice job on the views.

When making scenario outlines, don't make them so complex that you sacrifice readability.  It's difficult but important to strike a good balance between DRYing your code/tests and making them understandable.

Take a look at the comments and refactor.

**Pull request closed without merging**